pkg/controller/networkpolicy/tier.go

```diff
@@ -0,0 +1,165 @@
+	"k8s.io/klog"
+
+	secv1alpha1 "github.com/vmware-tanzu/antrea/pkg/apis/security/v1alpha1"
+)
+
+var (
+	// retryInitTier maintains the number of times Tier creation will be attempted
+	// for default Tiers during initialization.
+	retryInitTier     = 5
+	retryInitInterval = 2 * time.Second
+	// defaultTierPriority maintains the lowest priority for the system generated
+	// default Tier.
+	defaultTierPriority = uint32(250)
+	// priorityMap maintains the Tier priority associated with system generated
+	// Tier names.
+	priorityMap = map[string]uint32{
+		"application": defaultTierPriority,
+		"platform":    uint32(150),
+		"networkops":  uint32(100),
+		"securityops": uint32(50),
+		"emergency":   uint32(5),
+	}
+	// staticTierSet maintains the names of the static tiers such that they can
+	// be converted to corresponding Tier CRD names.
+	staticTierSet = sets.NewString("Emergency", "SecurityOps", "NetworkOps", "Platform", "Application")
+	// systemGeneratedTiers are the Tier CRs to be created at init.
+	systemGeneratedTiers = []*secv1alpha1.Tier{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "application",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["application"],
+				Description: "[READ-ONLY]: System generated default Application Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "platform",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["platform"],
+				Description: "[READ-ONLY]: System generated Platform Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "networkops",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["networkops"],
+				Description: "[READ-ONLY]: System generated NetworkOps Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "securityops",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["securityops"],
+				Description: "[READ-ONLY]: System generated SecurityOps Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "emergency",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["emergency"],
+				Description: "[READ-ONLY]: System generated Emergency Tier",
+			},
+		},
+	}
+)
+
+// InitializeTiers initializes the default Tiers created by Antrea on init. It
+// will first attempt to retrieve the Tier by it's name from K8s and if missing,
+// create the CR. InitializeTiers will be called as part of a Post-Start hook
+// of antrea-controller's APIServer.
+func (n *NetworkPolicyController) InitializeTiers() {
+	for _, t := range systemGeneratedTiers {
+		// Check if Tier is already present.
+		tr, err := n.tierLister.Get(t.Name)
+		if tr != nil && err == nil {
+			// Tier is already present.
+			klog.Infof("%s Tier already created", t.Name)
+			continue
+		}
+		err = n.initTier(t)
+		if err != nil {
+			// System generated Tiers were not initialized within an acceptable delay.
+			klog.Errorf("Failed to create system Tier %s on init", t.Name)
```

Comment:
The error doesn't fail the startup, then should it keep retrying with backoff until success so that it can recover itself? otherwise there will be no available tiers if the K8s service is not available when the controller starts up?

Reply:
done